### PR TITLE
Implement segfault handler in Crystal

### LINKED ISSUE #3839: Rewrite sigfault.c in Crystal
Right now we have a sigfault handler to print a nice error message on segmentation faults. It's located at [sigfault.c](https://github.com/crystal-lang/crystal/blob/d4ef40cd7ab4e7ea62141d9222e792c46af5ed5c/src/ext/sigfault.c)

This was written in C because if I remember correctly the internal structure of either `siginfo_t`, `stack_t` or `sigaction` were a bit complex, and even some properties are actually C macros.

The problem with this being written in C is that to compile the compiler we currently need to compile this file, so it's an extra step to remember. This is also kind of annoying when cross-compiling the compiler, because `sigfault.c` is compiled at the host machine, not the target machine. I think having this code re-written in Crystal will make things much simpler.

@ysbaddaden Do you think this can be done with https://github.com/ysbaddaden/posix ?

### PR DESCRIPTION
**Pull request by maxfierke**
Resolves #3839

Required adding some C bindings for various signal structures,
based on rust's fairly comprehensive libc bindings.

This removes the need to build and link a separate C library and
reduces the complexity of cross-compiling (at build-time), with
some added complexity when porting to a new platform.

`siginfo_t` can be fairly complex, involving large context-dependent
unions that vary across architecture and platform. In Crystal's
case, we only care about the segfault version of the structure,
so we omit the others.

Opening as draft first for feedback before more thoroughly verifying on all supported platforms. Tier 1 platforms verified so far. will take some time to get to the BSDs, etc.

- Tier I
  - [X] All verified (CI)
- Tier 2 (manual)
  - [x] DragonFly BSD
    - No available ports. Ran into issues cross-compiling (different LLVM versions available, libressl instead of openssl), but validated that `siginfo_t` bindings look correct
  - [X] FreeBSD (12.0)
    - std_spec hangs on io_spec but completes the segfault handler tests fine
  - [x] NetBSD
    - Ports has too old a version to compile Crystal master, but based on reading their libc source, `siginfo_t` bindings look correct.
  - [x] OpenBSD
    - Ports only has Crystal 0.30.1 which is too old to compile Crystal master, and ran into issues cross-compiling. However, can verify the `siginfo_t` bindings seem correct (at least on releases from summer 2017 forward)

### CODE REVIEW

File: src/signal.cr
Comment (straight-shoota) [inline src/signal.cr:142]:
```suggestion
  def self.setup_segfault_handler
```
See #10454

File: src/signal.cr
Comment (straight-shoota) [inline src/signal.cr:143]: Just to confirm: Previously there was no flag protection for `LibExt.setup_sigfault_handler
`, right? It's probably a good idea to add it, to avoid duplicate initialization :+1:

File: src/signal.cr
Comment (straight-shoota) [inline src/signal.cr:155]: You should be able to directly use a pointer to the struct member. That would be a more literal translation of the C implementation.

```suggestion
    LibC.sigemptyset(pointerof(action.sa_mask))
```

File: src/signal.cr
Comment (straight-shoota) [inline src/signal.cr:146]: Local variables don't need to be declared at the top in Crystal, so I'd move this line down right before `action` is used for the first time for more concise context.

File: src/signal.cr
Comment (straight-shoota) [inline src/signal.cr:151]: A simple `nil` would also do.
```suggestion
    LibC.sigaltstack(pointerof(altstack), nil)
```
I don't care either way, though.

File: src/signal.cr
Comment (straight-shoota) [inline src/signal.cr:129]: Do we need to keep `fun __crystal_sigfault_handler` around? It seems it was only necessary to export as a function for libcrystal to call it. Now we don't need that any more, the implementation can be moved completely into a Crystal method. Or directly to this proc here.

This could also be a follow-up refactoring.

**Review by straight-shoota — COMMENTED** [review]
Thanks, this looks really great and should surely help avoid some friction with distribution packages and cross-compiling.

**Comment by straight-shoota** [comment]
This requires an update to https://github.com/crystal-lang/distribution-scripts/ after merging.

File: src/signal.cr
Comment (maxfierke) [reply src/signal.cr:143]: yeah, there was none in the C extension. (Though by default initialization only happens once on the main thread, so it's probably not a huge risk, but why not do it? 😄 )

File: src/signal.cr
Comment (maxfierke) [reply src/signal.cr:155]: Unfortunately, that doesn't compile. Seems `pointerof` doesn't work with struct members like that.

File: src/signal.cr
Comment (maxfierke) [reply src/signal.cr:129]: Yeah directly in the proc would make sense to me. I'll do some looking to see if anything else is depending on `__crystal_sigfault_handler` (either within crystal or naughty libraries 😄 )

Comment (straight-shoota) [reply src/signal.cr:129]: I don't think the proc needs to be assigned to a class var either, it can just be initialized directly inside the setup method.

File: src/signal.cr
Comment (straight-shoota) [reply src/signal.cr:143]: Totally. Some user code could accidentally call `Signal.setup_sigfault_handle`.

File: src/signal.cr
Comment (straight-shoota) [reply src/signal.cr:155]: Sry, I missed the `@` for accessing an ivar: `pointerof(action.@sa_mask)` works.

File: src/signal.cr
Comment (maxfierke) [reply src/signal.cr:129]: My understand of `proc` callbacks passed to C is that we still need a reference to it on the class (or somewhere else in Crystal-land), otherwise it might be subject to GC?

Comment (straight-shoota) [reply src/signal.cr:129]: Oh, yeah right.

File: src/signal.cr
Comment (maxfierke) [reply src/signal.cr]: oh nice, that works!

**Review by straight-shoota — APPROVED** [review]

**Comment by maxfierke** [comment]
@straight-shoota is this something that could be slotted in for 1.1.0? I'd be happy to create a PR for the distribution-scripts work to support it, if that's something that's possible for someone outside the core team to do.

**Comment by straight-shoota** [comment]
We still need a second core team member to review and approve this PR.

Contributions to distribution-scripts are welcome 👍

File: src/signal.cr
Comment (asterite) [inline src/signal.cr:59]: Does this need to be exposed in Singal?

**Review by asterite — APPROVED** [review]
This is **GREAT**! I just left a small question.

File: src/signal.cr
Comment (Fryguy) [inline src/signal.cr:149]: Not for this PR, but would it be helpful to print the name of the signal?  In my experience the number is hard to understand.  Perhaps something like:

```suggestion
      Crystal::System.print_error "Invalid memory access (signal %d: SIG%s) at address 0x%lx\n", sig, Signal.new(sig), addr
```

File: src/signal.cr
Comment (maxfierke) [reply src/signal.cr:59]: I would guess probably not, unless there's a reason to expose stack size to users, but it's not clear to me whether we consider `LibC` bindings to be considered public API, and not clear whether `sigaltstack` and other functions for working with stacks would be relevant to most, if any, users. So I'll probably update this to avoid exposing `STKSZ` on `Signal`

Comment (asterite) [reply src/signal.cr:59]: `LibC` bindings are not public API. They are there to be used by the standard library. Users can use `LibC`, but in most cases they shouldn't need to.

So yes, for now let's not add it to Signal's public API. We could always add it later if needed.

Comment (maxfierke) [reply src/signal.cr]: great! addressed in 786c7d9

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**
